### PR TITLE
dirichlet noise added to prior probabilities during self play

### PR DESCRIPTION
**Pull request by jrbuhl93**
I don't completely understand the mathematics of the dirichlet distribution, but I arbitrarily chose an alpha of 0.6 as the default value for dirichletAlpha in main.py because that seems approximately right for othello. The one thing I am wondering about my code is that I am not sure when to apply it to the prior probabilities generated by the neural network for a board s. Currently I am applying it at all the search board nodes because otherwise only the first move would get the dirichlet noise due to the caching of the prior probabilities and policies.

But the paper seems to imply it should only be applied at the root node:

>  Additional exploration is achieved by adding Dirichlet noise to the prior probabilities in the root node s0, specifically P(s, a) = (1 − ε)p + εη, where η ∼ Dir(0.03) and ε = 0.25; this noise ensures that all moves may be tried, but the search may still overrule bad moves.

Thoughts and feedback are appreciated.

### CODE REVIEW

**Comment by NMO13** [comment]
Thanks for sharing your code!
Some remarks:

-  The paper states: "_Additional exploration is achieved by adding Dirichlet noise to the prior probabilities in the root node..._" So the noise should only be added to the root node
- You should only add noise to valid moves. So something more like ```np.random.dirichlet([self.args.dirichletAlpha] * np.count_nonzero(valids))```

**Comment by jrbuhl93** [comment]
I made edits to my code to add in your remarks. Let me know if you believe it is behaving properly now.

**Comment by jrbuhl93** [comment]
I was able to get a Connect4 agent playing perfectly as far as I have explored. It has 100% wins moving first in the evaluation loop. It also does not make any moves that would result in a loss or a draw when I play against it with the aid of https://connect4.gamesolver.org/. I used the updated dirichlet algorithm I made with NMO13's input. 

Other changes to achieve this:
-I added code that would randomly flip the board during self play like what the alpha go zero paper describes.
-I changed the neural network architecture to be the stack described in the alpha go zero paper, but I only used 6 residual blocks instead of 19 or 39 to save on computation.
-I added a function to convert the input to the neural network to be 6x7x2. The first 6x7 is for the current players pieces denoted by ones, and the second is for the second players pieces also denoted by ones. I did this because I believe overloading the network to detect both 1s and -1s may inhibit its ability to process a position. This also matches with the alpha zero paper having separate layers in the stack for each player's pieces, and will scale to games with multiple types of pieces which I plan to now implement.

My config was as follows:
'numIters': 30,
'numEps': 100, 
'tempThreshold': 10, 
'updateThreshold': 0.55, 
'maxlenOfQueue': 200000,
'numMCTSSims': 150, 
'arenaCompare': 6,
'cpuct': 1.25,
'dirichletAlpha': 1.4,

I don't know the exact iterations due to having to restart the code because of memory leaks in my current implementation of self play parallelization, but it was probably ~50.

**Comment by jrbuhl93** [comment]
I believe I spoke too soon. The second player has begun playing alternative moves, allowing the training to continue for new types of positions.

File: MCTS.py [hunk 1]
```diff
@@ -29,7 +30,8 @@ def getActionProb(self, canonicalBoard, temp=1):
                    proportional to Nsa[(s,a)]**(1./temp)
         """
         for i in range(self.args.numMCTSSims):
-            self.search(canonicalBoard)
+            dir_noise = (i == 0 and self.dirichlet_noise)
```
Comment (NMO13) [inline MCTS.py:33]: Not 100% sure, but I would remove the i == 0 and apply on every new MCTS run. That increases the exploration a bit more.

File: MCTS.py [hunk 2]
```diff
@@ -79,6 +81,8 @@ def search(self, canonicalBoard):
             self.Ps[s], v = self.nnet.predict(canonicalBoard)
             valids = self.game.getValidMoves(canonicalBoard, 1)
             self.Ps[s] = self.Ps[s]*valids      # masking invalid moves
+            if self.dirichlet_noise:
+                self.applyDirNoise(s, valids)
```
Comment (NMO13) [inline MCTS.py:85]: Why do you add noise at leave nodes?

File: MCTS.py [hunk 3]
```diff
@@ -96,6 +100,10 @@ def search(self, canonicalBoard):
             return -v
 
         valids = self.Vs[s]
+        if self.dirichlet_noise:
```
Comment (NMO13) [inline MCTS.py:103]: I think the self is wrong.

**Comment by NMO13** [comment]
I added some comments to the code.

> Other changes to achieve this:
-I added code that would randomly flip the board during self play like what the alpha go zero paper describes.
-I changed the neural network architecture to be the stack described in the alpha go zero paper, but I only used 6 residual blocks instead of 19 or 39 to save on computation.
-I added a function to convert the input to the neural network to be 6x7x2. The first 6x7 is for the current players pieces denoted by ones, and the second is for the second players pieces also denoted by ones. I did this because I believe overloading the network to detect both 1s and -1s may inhibit its ability to process a position. This matches with the alpha zero paper having separate layers in the stack for each player's pieces and will also scale to games with multiple types of pieces which I plan to now implement.
Sounds interesting but I would suggest that in a different PR bc. otherwise it gets too messy.

And please resolve the conflicts. ;)

File: MCTS.py (hunk 3 above)
Comment (jrbuhl93) [reply MCTS.py:103]: You are right. Ironically, the self.dirichlet_noise was making the dir_noise = (i == 0 and self.dirichlet_noise) in the earlier comment not do anything.

File: MCTS.py (hunk 1 above)
Comment (jrbuhl93) [reply MCTS.py:33]: I am not sure if the dir_noise should be applied more than once to s0. I would think this would cause too much drift from the original good probabilities. But due to my bug this is what was happening and it appeared to be working. Another more complicated option would be to still store the original s0 policy and apply a different dirichlet noise each simulation.

File: MCTS.py (hunk 2 above)
Comment (jrbuhl93) [reply MCTS.py:85]: I think this is fine as long as I am using dirichlet_noise and not self.dirichlet_noise. The reason for this is that the first move of the game will be a leaf node. This should be the only time dirichlet_noise is applied at a leaf node. I might verify this as a sanity check.

**Comment by jrbuhl93** [comment]
Thank you once again for your comments. I removed the bug (quite a large one) caused by calling self.dirichlet_noise instead of dirichlet_noise in the code I committed to my pull request. I also fixed this in the code for the model I am training now. I saw the results of your dirichlet_noise experiment, and am glad the evidence matches our intuitions. How similar is your code for implementing dirichlet noise?

Also let me know what you think about the i==0 and leaf node situations.

File: MCTS.py (hunk 1 above)
Comment (NMO13) [reply MCTS.py:33]: Thats the reason why I don't override the Ps. I add noise to them, calculate the ```u``` value and throw this value away again. Next run, I start with the original Ps.

Comment (NMO13) [reply MCTS.py:33]: Ok the  i == 0 makes sense now.

File: MCTS.py (hunk 2 above)
Comment (NMO13) [reply MCTS.py:85]: Makes sense.

**Comment by NMO13** [comment]
> How similar is your code for implementing dirichlet noise?

Pretty similar. The main difference is that I don't override Ps, as already mentioned. 
I do:
```p = np.copy(self.Ps[s])
        if self.add_noise and depth == 0:
            self.add_dirichlet(p, valids)
```
Notice the np.copy. And then I do:

```
if (s,a) in self.Qsa:
                    u = self.Qsa[(s,a)] + self.args.cpuct*p[a]*math.sqrt(self.Ns[s])/(1+self.Nsa[(s,a)])
                else:
                    u = self.args.cpuct*p[a]*math.sqrt(self.Ns[s] + EPS)     # Q = 0 ?
```
That doesn't drift the probabilities too far away.

**Comment by NMO13** [comment]
`I don't know the exact iterations due to having to restart the code because of memory leaks in my current implementation of self play parallelization, but it was probably ~75.`

Do you still have the weights of this agent? I would love to pit my version against yours.

**Comment by jrbuhl93** [comment]
> `I don't know the exact iterations due to having to restart the code because of memory leaks in my current implementation of self play parallelization, but it was probably ~75.`
> 
> Do you still have the weights of this agent? I would love to pit my version against yours.

I do! It is training right now! But I fear it would be a little complicated getting it running for you. I had to make some small changes to how nnets are loaded because of a bug in tensorflow2. Maybe we could set up a time for google hangout and manually input the opponents moves? I'll send you an email.

**Comment by jrbuhl93** [comment]
I think I like my implementation of dirichlet noise more, because I think it is more likely deep mind implemented it like I did based on my understanding of the paper. It seems to me to be better to search a path repeatedly in depth (based on if that path got a big increase in P from the dir noise). If you randomly assign dir noise for each simulation, I think the benefit  will be more like adding uniform noise because the variation of the noise will be less with the repeated sampling.

Do you agree or disagree?

**Comment by NMO13** [comment]
Ok so my inutition was to add more like unifrom noise like you already mentioned. Just to make sure that the other paths are not "starving". I was just scared that your solution of repeatedly overriding Ps could mess the probabilities too much up. But its difficult for me to evaluate and of course, I might be wrong and your solution works even better.

**Comment by NMO13** [comment]
I am currently training your version. I am curious how it will perform against my version.

**Comment by jrbuhl93** [comment]
> I am currently training your version. I am curious how it will perform against my version.

To minimize variables, you should use the dirichlet alpha you used to train your other model. I got very far with 1.4, but it got stuck for ~15 iterations on all ties. I just dropped the alpha to 0.9, so hopefully this is enough variation for it to find the optimal solution. What alpha did you use when you trained?

**Comment by NMO13** [comment]
I used 0.6. And for the first 100 iterations it finds a better model every ~4  iterations.
So here are some results. I let them pit against each other for 50 games.

Comparing your version vs mine:
(13, 24, 13) -> yours seems to work better

Comparing your version vs the original one:
(10, 32, 8) -> your version is much stronger

Conclusion: 
Of course, I just tested your algo for one game but I am highly confident that it works for other games equally well. 
You did well. I would merge your algorithm into the master. @suragnair what do you think?

File: MCTS.py [hunk 4]
```diff
@@ -96,6 +100,10 @@ def search(self, canonicalBoard):
             return -v
 
         valids = self.Vs[s]
+        if dirichlet_noise:
+            self.applyDirNoise(s, valids)
+            sum_Ps_s = np.sum(self.Ps[s])
+            self.Ps[s] /= sum_Ps_s      # renormalize
```
Comment (greghe) [inline MCTS.py:106]: `applyDirNoise` has already calculatesd a normalized sequence when it added two normalized sequences (scaled by 0.75 and 0.25, respectively), so there is no need to renormalize `Ps[s]` here – since doing so will have no effect
